toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -209,21 +292,149 @@ func resolveGraphNodes(dependencyGraph *pkggraph.PkgGraph, inputSummaryFile, out
+		return err
+	}
+
+	// Replace all '/' with '_' in the package name to get a valid timestamp name
+	// e.g. "bin/ls" -> "bin_ls"
+	tsName := strings.ReplaceAll(buildNode.VersionedPkg.Name, "/", "_")
+	timestamp.StartEvent(fmt.Sprintf("downloading delta node %s", tsName), nil)
+	defer timestamp.StopEvent(nil)
+
+	lookup, err = realDependencyGraph.FindExactPkgNodeFromPkg(buildNode.VersionedPkg)
+	if err != nil {
+		err = fmt.Errorf("can't find build node '%s' in graph: %w", buildNode, err)
+		return err
+	}
+	if lookup == nil || lookup.RunNode == nil {
+		err = fmt.Errorf("can't find run lookup '%v' in graph", lookup)
+		return err
+	}
+
+	runNode := lookup.RunNode
+
+	// Get the final output path for the build node if we don't convert it to a delta node
+	originalRpmPath := buildNode.RpmPath
+	foundFinalRPM, err := file.PathExists(originalRpmPath)
+	if err != nil {
+		return fmt.Errorf("can't check if final RPM '%s' exists: %w", originalRpmPath, err)
+	}
+
+	// Only download dependencies for delta RPMs if we don't already have the RPM in the out/RPMS folder
+	if !foundFinalRPM {
+		// We have the expected rpm path, take the base name and strip .rpm off to get a name we can pass to tdnf
+		// to download the delta RPM
+		// e.g. "/home/user/repo/out/RPMS/x86_64/pkg-1.0-1.cm2.x86_64.rpm" -> "pkg-1.0-1.cm2.x86_64"
+		fullyQualifiedRpmName := filepath.Base(originalRpmPath)
+		fullyQualifiedRpmName = strings.TrimSuffix(fullyQualifiedRpmName, ".rpm")
+
+		// Convert the name back into the expected path in the RPM cache. This is where the cloner is expected to put
+		// the RPM when it downloads it.
+		cachedRPMPath := rpmPackageToRPMPath(fullyQualifiedRpmName, deltaRpmDir)
+		foundCacheRPM, err := file.PathExists(cachedRPMPath)
+		if err != nil {
+			return fmt.Errorf("can't check if cached RPM '%s' exists: %w", cachedRPMPath, err)
+		}
+
+		// We will likely try to download the delta RPM multiple times across different nodes, so only do it if we don't
+		// already have it in the cache.
+		if !foundCacheRPM {
+			// Avoid any processing since we know the exact RPM we want to download
+			_, err = cloner.CloneRawPackageNames(downloadDependencies, fullyQualifiedRpmName)
+			if err != nil {
+				logger.Log.Warnf("Can't find delta RPM to download for %s: %s (local copy may be newer than published version)", fullyQualifiedRpmName, err)
+				return nil
+			}
+		}
+
+		foundCacheRPM, err = file.PathExists(cachedRPMPath)
+		if err != nil {
+			return fmt.Errorf("can't check if cached RPM '%s' exists: %w", cachedRPMPath, err)
+		}
+		if foundCacheRPM {
+			buildNode.State = pkggraph.StateDelta
+			runNode.State = pkggraph.StateDelta
+
+			// Update the build and run nodes to point to the new RPM in the cache
+			runNode.RpmPath = cachedRPMPath
+			buildNode.RpmPath = cachedRPMPath
+
+			logger.Log.Infof("Delta Progress %d%%: delta RPM found for '%s-%s'.", progress, buildNode.VersionedPkg.Name, buildNode.VersionedPkg.Version)
```

Comment:
>logger.Log.Infof("Delta Progress %d%%: delta RPM found for '%s-%s'.", progress, buildNode.VersionedPkg.Name, buildNode.VersionedPkg.Version) [](http://example.com/codeflow?start=3&length=140)

Similar request as with `resolveSingleNode` to have the caller report the progress instead of this function having to accept an additional `progress` argument, which is not related to its main function.

Reply:
This one is a bit harder since we have a few types of output we can generate... I'll see if I can rework this a bit.